tests/utils/test_data_utils.py

```diff
@@ -0,0 +1,51 @@
+# Copyright The FMS HF Tuning Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# SPDX-License-Identifier: Apache-2.0
+# https://spdx.dev/learn/handling-license-info/
+
+# First Party
+from tests.data import TWITTER_COMPLAINTS_DATA
+from tuning.utils import data_utils
+
+# Third Party
+import datasets
+import pytest
+
+def test_formatting_function():
+    json_dataset = datasets.load_dataset("json", data_files=TWITTER_COMPLAINTS_DATA)
+    template = "### Input: {{Tweet text}} \n\n ### Response: {{text_label}}"
+    # First response from the data file that is read.
+    expected_response = "### Input: @HMRCcustomers No this is my first job \n\n ### Response: no complaint"
+    formatted_dataset,  dataset_text_field = data_utils.formatting_function(json_dataset, template)
+    # a new dataset_text_field is created in Dataset
+    assert dataset_text_field in formatted_dataset['train'][0]
+    assert formatted_dataset['train'][0][dataset_text_field] == expected_response
+
+def test_formatting_function_adds_eos_token():
+    json_dataset = datasets.load_dataset("json", data_files=TWITTER_COMPLAINTS_DATA)
+    template = "### Input: {{Tweet text}} \n\n ### Response: {{text_label}}"
+    # First response from the data file that is read.
+    expected_response = "### Input: @HMRCcustomers No this is my first job \n\n ### Response: no complaintEOS"
```

Comment:
Why is it assumed to be okay that appending a raw EOS to the string will not change the tokenization, since this is operating on raw strings and not appending a token? I.e., in this kind of contrived example, just tokenizing the final token as `complaintEOS` instead of two tokens, the latter of which is EOS

Reply:
i agree its not ideal, but I just duplicated existing behavior in train() https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tuning/sft_trainer.py#L235 to this function to avoid looping and formatting dataset twice. Appended it same way as it was happening before - we can make update in future PRs to correct in both places